### PR TITLE
[CBRD-23266] lock-free freelist & transaction

### PR DESCRIPTION
**Pull request by ribram**
http://jira.cubrid.org/browse/CBRD-23266

Integrate lock-free transaction system into freelist.

**freelist::free_node**

- extends tran::reclaimable and reuses its m_retired_next member (replaces `get_freelist_next ()`)
- when reclaimed by transaction system, it is added to available list

**freelist**

- replace all `T *` pointers with `free_node *`
- add system/table dependency; freelist has ownership on its table
- add transactions to claim/retire; note: claim will not end transaction (this behavior may be subject to change in the future)
- replace compare_exchange_strong with compare_exchange_weak in loops (better overall performance)

**other**

- descriptor

  - constructor fixes initial value of m_tranid
  - add retire/reclaim count statistics

- table

  - add start_tran/end_tran to avoid get_descriptor () calls
  - add retire/reclaim count statistics

- unit test

  - update to new freelist design
  - passes

### CODE REVIEW

File: src/base/lockfree_transaction_reclaimable.hpp [hunk 1]
```diff
@@ -47,8 +47,12 @@ namespace lockfree
     class reclaimable_node
     {
       public:
-	reclaimable_node () = default;
-	virtual ~reclaimable_node () = 0;   // to force abstract class
+	reclaimable_node ()
```
Comment (razvanradulescu) [inline src/base/lockfree_transaction_reclaimable.hpp]: Is there other specialization than free_node for reclaimable_node ?
Do we need reclaimable_node abstract class ?
I am not sure the cost of accessing virtual table of reclaimable_node * instances from descriptor methods is to be ignored.

File: src/base/lockfree_freelist.hpp [hunk 2]
```diff
@@ -229,46 +276,53 @@ namespace lockfree
...
     // allocating directly into available list
-    while (t == NULL)
+    while (node == NULL)
       {
 	force_alloc_block ();
-	t = pop_from_available ();
+	node = pop_from_available ();
+      }
+    assert (node != NULL);
+    if (m_available_count-- == 0)
```
[the remark this answers is not in the document]
Comment (ribram) [reply src/base/lockfree_freelist.hpp]: m_available_count is size_t, it cannot be negative.

File: src/base/lockfree_freelist.hpp [hunk 3]
```diff
@@ -69,14 +84,42 @@ namespace lockfree
...
+      friend freelist;
+
+      void set_owner (freelist &m_freelist);
+
+      void set_freelist_next (free_node *next);
+      void reset_freelist_next (void);
+      free_node *get_freelist_next ();
+
+      freelist *m_owner;
+      T m_t;
```
[the remark this answers is not in the document]
Comment (ribram) [reply src/base/lockfree_freelist.hpp]: I am not sure about composition vs derivation :). freelist always does default construction. I am not sure we need or we can do custom construction.

File: src/base/lockfree_transaction_reclaimable.hpp (hunk 1 above)
Comment (ribram) [reply src/base/lockfree_transaction_reclaimable.hpp]: Not sure, I will make some performance tests comparing this implementation with previous.

File: src/base/lockfree_freelist.hpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (razvanradulescu) [reply src/base/lockfree_freelist.hpp]:
```suggestion
   assert (m_available_count > 0);
   m_available_count--;
```

[the remark this answers is not in the document]
Comment (ribram) [reply src/base/lockfree_freelist.hpp]: fixed

**Pull request merged**